arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Iterable.kt

```diff
@@ -283,153 +283,36 @@ public inline fun <B, C, D, E, F, G, H, I, J, K, L> Iterable<B>.zip(
 internal fun <T> Iterable<T>.collectionSizeOrDefault(default: Int): Int =
   if (this is Collection<*>) this.size else default
 
-@Deprecated("traverseEither is being renamed to traverse to simplify the Arrow API", ReplaceWith("traverse(f)", "arrow.core.traverse"))
-public inline fun <E, A, B> Iterable<A>.traverseEither(f: (A) -> Either<E, B>): Either<E, List<B>> =
-  traverse(f)
-
-@OptIn(ExperimentalTypeInference::class)
-@OverloadResolutionByLambdaReturnType
-public inline fun <E, A, B> Iterable<A>.traverse(f: (A) -> Either<E, B>): Either<E, List<B>> {
-  val destination = ArrayList<B>(collectionSizeOrDefault(10))
-  for (item in this) {
-    when (val res = f(item)) {
-      is Right -> destination.add(res.value)
-      is Left -> return res
-    }
-  }
-  return destination.right()
-}
-
-@Deprecated("sequenceEither is being renamed to sequence to simplify the Arrow API", ReplaceWith("sequence()", "arrow.core.sequence"))
-public fun <E, A> Iterable<Either<E, A>>.sequenceEither(): Either<E, List<A>> =
-  traverse(::identity)
-
-public fun <E, A> Iterable<Either<E, A>>.sequence(): Either<E, List<A>> =
-  traverse(::identity)
-
-@OptIn(ExperimentalTypeInference::class)
-@OverloadResolutionByLambdaReturnType
-public inline fun <A, B> Iterable<A>.traverse(f: (A) -> Result<B>): Result<List<B>> {
-  val destination = ArrayList<B>(collectionSizeOrDefault(10))
-  for (item in this) {
-    f(item).fold(destination::add) { throwable ->
-      return@traverse Result.failure(throwable)
-    }
-  }
-  return success(destination)
-}
+public inline fun <Error, A, B> Iterable<A>.mapAccumulating(
+  crossinline combine: (first: Error, second: Error) -> Error,
+  transform: (A) -> Either<Error, B>,
+): Either<Error, List<B>> =
+  fold<A, Either<Error, ArrayList<B>>>(Right(ArrayList(collectionSizeOrDefault(10)))) { acc, a ->
+    when (val res = transform(a)) {
+      is Right -> when (acc) {
+        is Right -> acc.also { acc.value.add(res.value) }
+        is Left -> acc
```

Comment:
@Atternatt looking at this implementation I think this is incorrect 🤔
The `combine` parameter is also unused, this looks like `traverseEither`.

Did we put all the relevant tests back?

Reply:
I think we're going to need something nasty like `EMPTY_VALUE` and a specialised `combine` for it to do this efficiently :/

Reply:
```kotlin
public inline fun <Error, A, B> Iterable<A>.mapAccumulating(
  combine: (first: Error, second: Error) -> Error,
  transform: (A) -> Either<Error, B>
): Either<Error, List<B>> {
  val right = ArrayList<B>(10)
  var left: Any? = EMPTY_VALUE
  forEach { a ->
    when(val res = transform(a)) {
      is Either.Left -> left = if (left == EMPTY_VALUE) res.value else combine(EMPTY_VALUE.unbox(left), res.value) 
      is Either.Right -> right.add(res.value)
    }
  }
  return if (left == EMPTY_VALUE) right.right() else EMPTY_VALUE.unbox(left)
}
```

I think this would be the most efficient implementation.